### PR TITLE
[IE CLDNN] Convolutition optimizations for small channels count

### PR DESCRIPTION
**Pull request by sshlyapn**

### CODE REVIEW

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_mmad_bfyx_to_b_fs_yx_fsv32.cpp [hunk 1]
```diff
@@ -122,24 +123,28 @@ ConvolutionKernel_mmad_bfyx_to_b_fs_yx_fsv32::AutoTuneOption ConvolutionKernel_m
...
     }
 
     return option;
 }
 
-static size_t get_slm_byte_size(const convolution_params &cp, size_t lws, size_t block_size) {
-    return (cp.stride.x * (lws * block_size - 1) + (cp.weights.X().v - 1) * cp.dilation.x + 1)*
-            cp.weights.Y().v * cp.weights.Z().v * sizeof(int32_t);
+static size_t get_slm_byte_size(const convolution_params &cp, size_t lws, size_t block_size_x, size_t block_size_y) {
+    const size_t input_y_height = cp.stride.y * (block_size_y - 1) + (cp.weights.Y().v - 1) * cp.dilation.x + 1;
```
Comment (lznamens) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_mmad_bfyx_to_b_fs_yx_fsv32.cpp:139]: cp.dilation.y

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_mmad_bfyx_to_b_fs_yx_fsv32.cpp [hunk 2]
```diff
@@ -184,37 +189,42 @@ JitConstants ConvolutionKernel_mmad_bfyx_to_b_fs_yx_fsv32::GetJitConstants(const
...
     jit.Merge(MakeTypeJitConstants(GetPackedInputType(params), "PACKED_IN"));
     jit.Merge(MakeTypeJitConstants(GetPackedType(params.output.GetDType(), 2), "PACKED_OUT"));
 
     if (!params.fused_ops.empty()) {
         auto input_dt = GetActivationType(params);
         if (WeightsTensor::ChannelsCount(GetPreferredWeightsLayout(params)) == 5) {
-            FusedOpsConfiguration conf0 = {"_0", {"b", "(fg*32 + lid)", "z", "y", "(x+i)"}, "res0", input_dt, 1};
-            FusedOpsConfiguration conf1 = {"_1", {"b", "(fg*32 + lid+16)", "z", "y", "(x+i)"}, "res1", input_dt, 1};
+            FusedOpsConfiguration conf0 = {"_0", {"b", "(fg*32 + lid)", "z", "(y+j)", "(x+i)"}, "res0", input_dt, 1};
+            FusedOpsConfiguration conf1 = {"_1", {"b", "(fg*32 + lid+16)", "z", "(y+j)", "(x+i)"}, "res1", input_dt, 1};
```
Comment (lznamens) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_mmad_bfyx_to_b_fs_yx_fsv32.cpp:218]: Please add spaces:
`+ lid + 16`

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_mmad_bfyx_to_b_fs_yx_fsv32.cpp [hunk 3]
```diff
@@ -184,37 +189,42 @@ JitConstants ConvolutionKernel_mmad_bfyx_to_b_fs_yx_fsv32::GetJitConstants(const
...
             jit.Merge(MakeFusedOpsJitConstants(params, {conf0, conf1}));
         } else {
             if (GetPreferredWeightsLayout(params) == WeightsLayout::os_is_yx_osv32_isv4) {
-                FusedOpsConfiguration conf0 = {"_0", {"b", "(fg*32 + lid)", "y", "(x+i)"}, "res0", input_dt, 1};
-                FusedOpsConfiguration conf1 = {"_1", {"b", "(fg*32 + lid+16)", "y", "(x+i)"}, "res1", input_dt, 1};
+                FusedOpsConfiguration conf0 = {"_0", {"b", "(fg*32 + lid)", "(y+j)", "(x+i)"}, "res0", input_dt, 1};
+                FusedOpsConfiguration conf1 = {"_1", {"b", "(fg*32 + lid+16)", "(y+j)", "(x+i)"}, "res1", input_dt, 1};
```
Comment (lznamens) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_mmad_bfyx_to_b_fs_yx_fsv32.cpp:223]: The same as in comment above.

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_mmad_bfyx_to_b_fs_yx_fsv32.cl [hunk 4]
```diff
@@ -114,7 +115,7 @@ KERNEL(convolution_mmad_bfyx_to_b_fs_yx_fsv32)(
     const int input_y = y * STRIDE_SIZE_Y - PADDING_SIZE_Y;
     const int input_z = z * STRIDE_SIZE_Z - PADDING_SIZE_Z;
 
-    ACCUMULATOR_TYPE_VEC acc[2] = { 0 }; // 2*16 packed channels * OUTPUT_X_BLOCK_SIZE
+    ACCUMULATOR_TYPE_VEC acc[2 * OUTPUT_Y_BLOCK_SIZE] = { 0 }; // 2*16 packed channels * OUTPUT_X_BLOCK_SIZE
```
Comment (lznamens) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_mmad_bfyx_to_b_fs_yx_fsv32.cl:118]: I suppose the comment isn't relevant now.

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_mmad_bfyx_to_b_fs_yx_fsv32.cl [hunk 5]
```diff
@@ -239,17 +239,34 @@ KERNEL(convolution_mmad_bfyx_to_b_fs_yx_fsv32)(
...
+#if !ASYMMETRIC_DATA_QUANTIZATION
+                if (y_cross_fm) {
+                    for (int xb = 0; xb < INPUT_LINE_SIZE; xb++) {
+                        line_cache[yb * INPUT_LINE_SIZE + xb] = 0;
+                    }
+                    continue;
+                }
+#endif
+                for (int xb = 0; xb < INPUT_LINE_SIZE; xb++) {
+                    line_cache[yb * INPUT_LINE_SIZE + xb] = slm[kd*SLM_LINE_SIZE*INPUT_Y_HEIGHT + ((kh*DILATION_SIZE_Y)+yb*STRIDE_SIZE_Y)*SLM_LINE_SIZE + sg*OUTPUT_X_BLOCK_SIZE*STRIDE_SIZE_X + xb];
```
Comment (lznamens) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_mmad_bfyx_to_b_fs_yx_fsv32.cl:257]: Too long line.

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_mmad_bfyx_to_b_fs_yx_fsv32.cl [hunk 6]
```diff
@@ -239,17 +239,34 @@ KERNEL(convolution_mmad_bfyx_to_b_fs_yx_fsv32)(
         if (z_cross_fm)
             continue;
 #endif
+        PACKED_IN_TYPE line_cache[INPUT_LINE_SIZE * OUTPUT_Y_BLOCK_SIZE];
         __attribute__((opencl_unroll_hint(FILTER_SIZE_Y)))
         for (int kh = 0; kh < FILTER_SIZE_Y ; ++kh) {
+#if OUTPUT_Y_BLOCK_SIZE > 1
+            for (int yb = 0; yb < OUTPUT_Y_BLOCK_SIZE; yb++) {
+                bool y_cross_fm = input_y + ((kh*DILATION_SIZE_Y)+yb*STRIDE_SIZE_Y) < 0 || input_y + ((kh*DILATION_SIZE_Y)+yb*STRIDE_SIZE_Y) >= INPUT0_SIZE_Y;
```
Comment (lznamens) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_mmad_bfyx_to_b_fs_yx_fsv32.cl:247]: I think you should replace this line under define`ASYMMETRIC_DATA_QUANTIZATION`. And the same in the code line 261.

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_mmad_bfyx_to_b_fs_yx_fsv32.cl [hunk 7]
```diff
@@ -259,21 +276,32 @@ KERNEL(convolution_mmad_bfyx_to_b_fs_yx_fsv32)(
...
-                    // src[i] = slm[kh*SLM_LINE_SIZE + (sg*OUTPUT_X_BLOCK_SIZE + i)*STRIDE_SIZE_X + kw*DILATION_SIZE_X];
-                    src[i] = line_cache[kw*DILATION_SIZE_X + STRIDE_SIZE_X*i];
-                    acc[0][i] = IMAD(acc[0][i], AS_INPUT0_TYPE_4(src[i]), as_char4(weights_data0));
-                    acc[1][i] = IMAD(acc[1][i], AS_INPUT0_TYPE_4(src[i]), as_char4(weights_data1));
-
+#if OUTPUT_Y_BLOCK_SIZE > 1
+                __attribute__((opencl_unroll_hint(OUTPUT_Y_BLOCK_SIZE)))
+                for (int yb = 0; yb < OUTPUT_Y_BLOCK_SIZE; yb++) {
+#else
+                #define yb 0
```
Comment (lznamens) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_mmad_bfyx_to_b_fs_yx_fsv32.cl:288]: Does it really improve something?

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_mmad_bfyx_to_b_fs_yx_fsv32.cl [hunk 8]
```diff
@@ -259,21 +276,32 @@ KERNEL(convolution_mmad_bfyx_to_b_fs_yx_fsv32)(
...
+#endif
+                    __attribute__((opencl_unroll_hint(OUTPUT_X_BLOCK_SIZE)))
+                    for (int i = 0; i < OUTPUT_X_BLOCK_SIZE; i++) {
+                        // src[i] = slm[kh*SLM_LINE_SIZE + (sg*OUTPUT_X_BLOCK_SIZE + i)*STRIDE_SIZE_X + kw*DILATION_SIZE_X];
```
Comment (lznamens) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_mmad_bfyx_to_b_fs_yx_fsv32.cl:292]: Please delete unused comment.

**Review by lznamens — CHANGES_REQUESTED** [review]
Also please update license info in *.cl-file.

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_mmad_bfyx_to_b_fs_yx_fsv32.cl (hunk 7 above)
Comment (sshlyapn) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_mmad_bfyx_to_b_fs_yx_fsv32.cl:289]: I was surprised, but yes, it really make sense in this case

**Review by lznamens — APPROVED** [review]
Thanks. LGTM.

**Review by vladimir-paramuzov — APPROVED** [review]

**Comment by vladimir-paramuzov** [comment]
@sshlyapn Could you please rebase onto the latest master? For some reason github doesn't want to merge this PR.

**Pull request merged**